src/clients/c++/perf_client/load_manager.cc

```diff
@@ -213,33 +306,297 @@ LoadManager::InitManagerInputs(
+                  std::to_string(batch1_bytesize) +
+                  " bytes for each batch, but provided data has " +
+                  std::to_string(it->second.size()) + " bytes");
+        }
+        data = (const uint8_t*)&(it->second)[0];
+      } else if (input_buf_.size() != 0) {
+        if (batch1_bytesize > input_buf_.size()) {
+          return nic::Error(
+              ni::RequestStatusCode::INTERNAL,
+              "input '" + input->Name() + "' requires " +
+                  std::to_string(batch1_bytesize) +
+                  " bytes for each batch, but generated data has " +
+                  std::to_string(input_buf_.size()) + " bytes");
+        }
+        data = &input_buf_[0];
+      } else {
+        return nic::Error(
+            ni::RequestStatusCode::INVALID_ARG,
+            "unable to find data for input '" + input->Name() + "'.");
+      }
+    } else {
+      size_t batch1_num_strings = 1;
+      if (!input->Shape().empty()) {
+        for (const auto dim : input->Shape()) {
+          batch1_num_strings *= dim;
+        }
+      } else {
+        for (const auto dim : input->Dims()) {
+          batch1_num_strings *= dim;
+        }
+      }
+
+      bool used_new_allocation = false;
+      std::vector<std::string>* string_data;
+
+      // if available, use provided data instead
+      auto it = input_string_data_.find(input->Name());
+      if (it != input_string_data_.end()) {
+        if (it->second.size() != batch1_num_strings) {
+          return nic::Error(
+              ni::RequestStatusCode::INVALID_ARG,
+              "input '" + input->Name() + "' requires " +
+                  std::to_string(batch1_num_strings) +
+                  " strings for each batch, but provided data has " +
+                  std::to_string(it->second.size()) + " strings.");
+        }
+        string_data = &it->second;
+      } else if (input_string_buf_.size() != 0) {
+        if (batch1_num_strings > input_string_buf_.size()) {
+          return nic::Error(
+              ni::RequestStatusCode::INTERNAL,
+              "input '" + input->Name() + "' requires " +
+                  std::to_string(batch1_num_strings) +
+                  " strings for each batch, but generated data has " +
+                  std::to_string(input_string_buf_.size()) + " strings.");
+        } else {
+          used_new_allocation = true;
+          string_data = new std::vector<std::string>(
+              input_string_buf_.begin(),
+              input_string_buf_.begin() + batch1_num_strings);
+        }
+      } else {
+        return nic::Error(
+            ni::RequestStatusCode::INVALID_ARG,
+            "unable to find data for input '" + input->Name() + "'.");
+      }
+
+      SerializeStringTensor(*string_data, &serialized_data);
+      batch1_bytesize = serialized_data.size();
+      data = (uint8_t*)&serialized_data[0];
+
+      if (used_new_allocation) {
+        delete string_data;
+      }
+    }
```

Comment:
The above is similar to what we do for preparing inputs for none shared memory case, can be extracted as a utility function?

Reply:
Done